### PR TITLE
Allow etcdctl to use a yaml config file

### PR DESCRIPTION
**Pull request by seeker89**
## TL;DR

Recently I've been working a lot with `etcdctl`, and it started bothering me a lot that it doesn't support a config file, like [etcd itself does](https://github.com/etcd-io/etcd/blob/release-3.4/etcd.conf.yml.sample). I ended up generating bash wrappers with the env vars baked in, which is largely suboptimal.

## How it works

By default, it looks for `~/.etcdctl/config.{yml,ini,toml}` config file. Let's say I have one here

```sh
$ cat ~/.etcdctl/config.yml 
endpoints: https://node-1:2379,https://node-2:2379,https://node-3:2379,https://node-4:2379,https://node-5:2379
cert: ./client.crt
key: ./client.key
cacert: ./ca.crt

$ ./bin/etcdctl endpoint health
https://node-3:2379 is healthy: successfully committed proposal: took = 41.935928ms
https://node-2:2379 is healthy: successfully committed proposal: took = 41.154774ms
https://node-4:2379 is healthy: successfully committed proposal: took = 44.257905ms
https://node-5:2379 is healthy: successfully committed proposal: took = 44.746235ms
https://node-1:2379 is healthy: successfully committed proposal: took = 44.33043ms
```

It also supports a new flag `--config` to explicitly specify the config file:

```sh
...
OPTIONS:
...
  -c, --config=""                               yaml config file with global flags
...
```

So you can specify a local file like this:

```sh
$ cat ./config.yml 
endpoints: https://node-1:2379,https://node-2:2379
cert: ./client.crt
key: ./client.key
cacert: ./ca.crt

$ ./bin/etcdctl endpoint health -c ./config.yml 
https://node-2:2379 is healthy: successfully committed proposal: took = 42.584739ms
https://node-1:2379 is healthy: successfully committed proposal: took = 44.27894ms
```

And to test things out, we can delete the default location and see it break:

```sh
$ rm ~/.etcdctl/config.yml 
$ ./bin/etcdctl endpoint health 
{"level":"warn","ts":"2021-04-23T17:08:32.431+0100","logger":"etcd-client","caller":"v3/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0xc0003b0700/#initially=[127.0.0.1:2379]","attempt":0,"error":"rpc error: code = DeadlineExceeded desc = latest balancer error: last connection error: connection error: desc = \"transport: Error while dialing dial tcp 127.0.0.1:2379: connect: connection refused\""}
127.0.0.1:2379 is unhealthy: failed to commit proposal: context deadline exceeded
Error: unhealthy cluster
```

### What about the env vars?

Doing it this way behaves exactly like if you specified flags on the command line. So you get the familar message about shadowing, if you specify both:

```sh
$ ETCDCTL_ENDPOINTS=https://node-1:2379  ./bin/etcdctl endpoint health
{"level":"fatal","ts":1619194014.967419,"caller":"flags/flag.go:85","msg":"conflicting environment variable is shadowed by corresponding command-line flag (either unset environment variable or disable flag)...
```

### What the binary size?

Finally, it looks like the `etcdctl` size went from 24 to 26MB (Darwin).

Thanks and let me know if this works for you.

### CODE REVIEW

**Comment by seeker89** [comment]
HI, would anyone like to take a look at this?

I'm not sure why [one of the Travis jobs seems to fail](https://travis-ci.com/github/etcd-io/etcd/jobs/500774650) with this error:

```sh
modularized licenses do not match given bill of materials
FAIL: 'bom' failed at Fri Apr 23 17:49:08 UTC 202
```

The only new dependency is `github.com/spf13/viper` and that's MIT licence, and it seems to be correctly detected:

```sh
> 	{
398a471,479
> 		"project": "github.com/spf13/viper",
> 		"licenses": [
> 			{
> 				"type": "MIT License",
> 				"confidence": 1
> 			}
> 		]
> 	},
```

File: etcdctl/ctlv3/ctl.go [hunk 1]
```diff
@@ -42,10 +45,14 @@ var (
...
 		Short:      cliDescription,
 		SuggestFor: []string{"etcdctl"},
+		PersistentPreRun: func(cmd *cobra.Command, args []string) {
+			HandleConfigFile(cmd)
+		},
 	}
 )
 
 func init() {
+	rootCmd.PersistentFlags().StringVarP(&globalFlags.ConfigFile, "config", "c", "", "yaml config file with global flags")
```
Comment (hexfusion) [inline etcdctl/ctlv3/ctl.go:55]: server uses config-file naming we should probably align?

Comment (hexfusion) [reply etcdctl/ctlv3/ctl.go:55]: consider aligning with server

```
Path to the client configuration file. Note that if a configuration file is provided, other command line flags and environment variables will be ignored.
```

**Review by hexfusion — COMMENTED** [review]
I don't see a lot of value with this besides the fact that the server has support. In general, ENV is very popular with any container based usage. I guess any additional use-case would help us understand why this would be needed.

> {"level":"fatal","ts":1619194014.967419,"caller":"flags/flag.go:85","msg":"conflicting environment variable is shadowed by corresponding command-line flag (either unset environment variable or disable flag)...

Sort of confusing as there is no flag but I suppose this is how the server responds?

**Comment by ptabor** [comment]
> The only new dependency is `github.com/spf13/viper` and that's MIT licence, and it seems to be correctly detected:

`./scripts/fix.sh` or `./scripts/updatebom.sh` should solve the issue (update the `bill-of-materials.json` file)

File: etcdctl/ctlv3/ctl.go (hunk 1 above)
Comment (seeker89) [reply etcdctl/ctlv3/ctl.go]: Changed to `config-file`. I thought that leaving the existing precedence model was better, but I'm open to suggestions.

**Comment by seeker89** [comment]
> > The only new dependency is `github.com/spf13/viper` and that's MIT licence, and it seems to be correctly detected:
> 
> `./scripts/fix.sh` or `./scripts/updatebom.sh` should solve the issue (update the `bill-of-materials.json` file)

Thanks! This seems to have worked 👍

**Comment by seeker89** [comment]
> I don't see a lot of value with this besides the fact that the server has support. In general, ENV is very popular with any container based usage. I guess any additional use-case would help us understand why this would be needed.
> 
> > {"level":"fatal","ts":1619194014.967419,"caller":"flags/flag.go:85","msg":"conflicting environment variable is shadowed by corresponding command-line flag (either unset environment variable or disable flag)...
> 
> Sort of confusing as there is no flag but I suppose this is how the server responds?

The main use case is for managing a farm of etcd clusters, and switching between contexts. ENV works fine, but to change from one cluster to another, you need a new set of environment variables, which leads to writing wrapper scripts that call `etcdctl` with different environment variables.

**Comment by hexfusion** [comment]
> > I don't see a lot of value with this besides the fact that the server has support. In general, ENV is very popular with any container based usage. I guess any additional use-case would help us understand why this would be needed.
> > > {"level":"fatal","ts":1619194014.967419,"caller":"flags/flag.go:85","msg":"conflicting environment variable is shadowed by corresponding command-line flag (either unset environment variable or disable flag)...
> > 
> > 
> > Sort of confusing as there is no flag but I suppose this is how the server responds?
> 
> The main use case is for managing a farm of etcd clusters, and switching between contexts. ENV works fine, but to change from one cluster to another, you need a new set of environment variables, which leads to writing wrapper scripts that call `etcdctl` with different environment variables.

I appreciate the context here and this does make sense.

**Comment by seeker89** [comment]
@hexfusion thanks!

Now, another test [started failing](https://travis-ci.com/github/etcd-io/etcd/jobs/503454652) (`TARGET=linux-amd64-integration-4-cpu`), although the other versions (1 and 2 cpu) seem to work? I don't think this change affected any of those, are the tests a little flaky? I'm seeing:

```sh
{"level":"warn","ts":"2021-05-06T16:38:30.768Z","logger":"etcd-client","caller":"v3/maintenance.go:221","msg":"failed to receive from snapshot stream; closing","error":"rpc error: code = Canceled desc = context canceled"}
{"level":"info","ts":"2021-05-06T16:38:30.769Z","logger":"etcd-client","caller":"v3/maintenance.go:211","msg":"opened snapshot stream; downloading"}
{"level":"warn","ts":"2021-05-06T16:38:31.774Z","logger":"etcd-client","caller":"v3/maintenance.go:221","msg":"failed to receive from snapshot stream; closing","error":"rpc error: code = DeadlineExceeded desc = context deadline exceeded"}
{"level":"warn","ts":"2021-05-06T16:38:52.646Z","logger":"etcd-client","caller":"v3/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0xc0006bc1c0/#initially=[unix://localhost:m00]","attempt":0,"error":"rpc error: code = InvalidArgument desc = etcdserver: authentication failed, invalid user ID or password"}
{"level":"warn","ts":"2021-05-06T16:38:52.676Z","logger":"etcd-client","caller":"v3/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0xc000240000/#initially=[unix://localhost:m00]","attempt":0,"error":"rpc error: code = InvalidArgument desc = etcdserver: authentication failed, invalid user ID or password"}
{"level":"warn","ts":"2021-05-06T16:38:53.026Z","logger":"etcd-client","caller":"v3/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0xc000240700/#initially=[unix://localhost:m10]","attempt":0,"error":"rpc error: code = FailedPrecondition desc = etcdserver: authentication is not enabled"}
FAIL
FAIL	go.etcd.io/etcd/tests/v3/integration/clientv3	122.163s
```

File: etcdctl/ctlv3/ctl.go [hunk 2]
```diff
@@ -42,10 +45,14 @@ var (
...
 		Short:      cliDescription,
 		SuggestFor: []string{"etcdctl"},
+		PersistentPreRun: func(cmd *cobra.Command, args []string) {
+			HandleConfigFile(cmd)
+		},
 	}
 )
 
 func init() {
+	rootCmd.PersistentFlags().StringVarP(&globalFlags.ConfigFile, "config-file", "c", "", "Path to the client configuration file. Note that other command line flags and environment variables take precedence.")
```
Comment (xiang90) [inline etcdctl/ctlv3/ctl.go:55]: `Path to the configuration files for etcdctl. Note that explicitly provided command-line flags and environment variables take precedence.`

File: etcdctl/ctlv3/ctl.go [hunk 3]
```diff
@@ -95,6 +102,27 @@ func init() {
...
+func HandleConfigFile(cmd *cobra.Command) {
+	v := viper.New()
+	if globalFlags.ConfigFile != "" {
+		v.SetConfigFile(globalFlags.ConfigFile)
+	} else {
+		v.SetConfigName("config")
+		v.AddConfigPath("$HOME/.etcdctl")
+	}
+	if err := v.ReadInConfig(); err != nil {
+		// Error, unless it's just a missing file
```
Comment (xiang90) [inline etcdctl/ctlv3/ctl.go:114]: Remove this comment line? It seems to be unnecessary. The code is self-explain.

Comment (seeker89) [reply etcdctl/ctlv3/ctl.go:114]: Sure

**Comment by hexfusion** [comment]
Can we rebase this into a single commit for the feature and a second commit updating the changelog ?

**Comment by seeker89** [comment]
@hexfusion squashed the commits and added another one for the changelog. I wasn't 100% sure which tree this should go to, so I attached the change to 3.4.16 that appears to be imminent?

**Comment by hexfusion** [comment]
>  I wasn't 100% sure which tree this should go to, so I attached the change to 3.4.16 that appears to be imminent?

This feature would be part of 3.5.0

**Comment by seeker89** [comment]
@hexfusion moved the changelog to 3.5.0

**Comment by seeker89** [comment]
@hexfusion @xiang90 @ptabor anything else you'd like changed, or is this good to go? Things become stale quite quickly in here :)

**Comment by hexfusion** [comment]
@seeker89 please clear merge conflicts

**Comment by seeker89** [comment]
@hexfusion done :)

**Comment by seeker89** [comment]
Ping @hexfusion

File: CHANGELOG-3.5.md [hunk 4]
```diff
@@ -227,6 +227,7 @@ Note that any `etcd_debugging_*` metrics are experimental and subject to change.
 - Add [`etcdctl get --count-only`](https://github.com/etcd-io/etcd/pull/11743) flag for output type `fields`.
 - Add [`etcdctl member list -w=json --hex`](https://github.com/etcd-io/etcd/pull/11812) flag to print memberListResponse in hex format json.
 - Changed [`etcdctl lock <lockname> exec-command`](https://github.com/etcd-io/etcd/pull/12829) to return exit code of exec-command.
+- Add support for a config file. Add [`etcdctl --config-file`](https://github.com/etcd-io/etcd/pull/12891) to specify a config file and search in default location in `$HOME/.etcdctl/config.yml`.
```
Comment (spzala) [inline CHANGELOG-3.5.md:230]: I would suggest a small re-write here. Make it explicit here by saying something like `Add support for providing etcdctl arguments in a config file` instead of `Add support for a config file.` to avoid any confusing with server config file in the first sentence before use read further. Also, it uses provided file or one from default location so may be something like "Add [`etcdctl --config-file`](https://github.com/etcd-io/etcd/pull/12891) flag to specify a config file or use `$HOME/.etcdctl/config.yml` if provided."

File: etcdctl/ctlv3/ctl.go [hunk 5]
```diff
@@ -96,6 +103,26 @@ func init() {
 	)
 }
 
+func HandleConfigFile(cmd *cobra.Command) {
+	v := viper.New()
+	if globalFlags.ConfigFile != "" {
+		v.SetConfigFile(globalFlags.ConfigFile)
+	} else {
+		v.SetConfigName("config")
```
Comment (spzala) [inline etcdctl/ctlv3/ctl.go:111]: I believe we need to explicitly set yaml as `v.SetConfigType("yaml")`

File: etcdctl/ctlv3/ctl.go [hunk 6]
```diff
@@ -44,10 +47,14 @@ var (
...
 		Short:      cliDescription,
 		SuggestFor: []string{"etcdctl"},
+		PersistentPreRun: func(cmd *cobra.Command, args []string) {
+			HandleConfigFile(cmd)
+		},
 	}
 )
 
 func init() {
+	rootCmd.PersistentFlags().StringVarP(&globalFlags.ConfigFile, "config-file", "c", "", "Path to the configuration file for etcdctl. Note that explicitly provided command-line flags and environment variables take precedence.")
```
Comment (spzala) [inline etcdctl/ctlv3/ctl.go:57]: So if a config file is provided or found one in the default location, any other provided command line flags and environment variables be ignored right ? Keeping something inline with what we already have for handling config file for server would be good -  https://github.com/etcd-io/etcd/blob/main/server/etcdmain/config.go#L136. Should also cover that if path provided then default config file will be ignored as well ?

**Review by spzala — COMMENTED** [review]
@sakateka few inline comments. Also, we should add this in the documentation for the etcdctl - https://github.com/etcd-io/etcd/tree/main/etcdctl

**Comment by sakateka** [comment]
- @sakateka 
+@seeker89
